toolchain/sem_ir/type_iterator.cpp

```diff
@@ -0,0 +1,226 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/type_iterator.h"
+
+#include <utility>
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/constant.h"
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::SemIR {
+
+auto TypeIterator::Next() -> Step {
+  while (!work_list_.empty()) {
+    auto next = work_list_.back();
+    work_list_.pop_back();
+
+    if (std::holds_alternative<EndType>(next)) {
+      return Step::End();
+    }
+
+    if (const auto* interface = std::get_if<SemIR::SpecificInterface>(&next)) {
+      auto args = GetSpecificArgs(interface->specific_id);
+      if (args.empty()) {
+        return Step::StartOnly(
+            Step::InterfaceStart{.interface_id = interface->interface_id});
+      } else {
+        Push(EndType());
+        PushArgs(args);
+        return Step::StartWithEnd(
+            Step::InterfaceStart{.interface_id = interface->interface_id});
+      }
+    }
+
+    if (const auto* symbolic = std::get_if<SymbolicType>(&next)) {
+      return Step::SymbolicType{.facet_type_id = symbolic->facet_type_id};
+    }
+
+    if (std::holds_alternative<NonTypeValue>(next)) {
+      return Step::Value();
+    }
+
+    SemIR::TypeId type_id = std::get<SemIR::TypeId>(next);
+    auto inst_id = sem_ir_->types().GetInstId(type_id);
+    auto inst = sem_ir_->insts().Get(inst_id);
+    CARBON_KIND_SWITCH(inst) {
```

Comment:
Not sure how you feel about moving this inline (it seems unused outside the switch):
```suggestion
    CARBON_KIND_SWITCH(sem_ir_->insts().Get(inst_id)) {
```

Note, generally I associate keeping the `inst` separate with one of the `::Kind` cases accessing it (e.g., for an `Any` type use).

Reply:
I could go either way, sometimes I do like to inline things into the switch. Here, it feels more clear leaving it outside for whatever reason - I like being able to see the name (and type in an ide) as a hint when reading this, and it's a useful value as you say that future code may want to use.